File: 145. Binary Tree Postorder Traversal/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        stack<TreeNode *> sta;
        vector<int> result;
        TreeNode *node = root;
        while (node != nullptr || !sta.empty()) {
            if (node != nullptr) {
                sta.push(node);
                result.push_back(node->val);
                node = node->right;
            } else {
                node = sta.top();
                sta.pop();
                node = node->left;
            }
        }
        reverse(result.rbegin(), result.rend());
        return result;
    }
};

File: 145. Binary Tree Postorder Traversal/Python/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack = []
        result = []
        node = root
        while node or stack:
            if node:
                stack.append(node)
                result.append(node.val)
                node = node.right
            else:
                node = stack.pop()
                node = node.left
        result.reverse()
        return result
